031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,161 @@
+# TLS encrypting the Kafka Connect REST API
+
+This proposal adds the option to encrypt communication with the Kafka Connect REST API by enabling TLS encryption on the Kafka Connect REST listener.
+
+## Current situation
+
+Currently, instances of Kafka Connect that are deployed by the Strimzi operators are configured with the default REST API endpoint settings.
+This means that the Kafka Connect REST API endpoint uses HTTP on port 8083 and that the Strimzi `KafkaConnect`, `KafkaConnector` and `KafkaMirrorMaker2` operators make unsecured REST client calls based on this default configuration.
+
+The default network policies created by the Strimzi operators restrict incoming REST API calls to only allow access from the operator pod.
+Users can define further network policies to override the default policy and allow wider access.
+
+## Motivation
+
+There is currently no way to TLS encrypt the Kafka Connect REST API, leaving the Connect REST listener as perhaps the last key endpoint that cannot currently be encrypted using Strimzi.
+
+## Proposal
+
+This proposal adds REST api configuration options to the `KafkaMirrorMaker2` and `KafkaConnect` CRDs.
+
+A new field is added to the `KafkaConnect` and `KafkaMirrorMaker2` specs named `restListeners`, which allows users to configure the Kafka Connect configuration to have:
+  - a single HTTP listener on port 8083
+  - a single encrypted HTTPS listener on port 8443
+  - both an HTTP listener on port 8083 and an encrypted HTTPS listener on port 8443
+  
+This enables users to access the REST API with TLS if required.
+The `KafkaConnector`, `KafkaConnect` and `KafkaMirrorMaker2` operators will use the TLS encrypted listener when it is enabled, even if an unencrypted listener is present.
+
+The `restListeners` field is designed for potential future extension, and is a list of items that contains a `protocol` field and fields for providing certificates if required `certChainAndKey`, `strimziTruststoreCert`
+
+For this proposal, the `protocol` field only supports two values: `http` and `https`. If the `restListeners` list is empty the `KafkaConnect` and `KafkaMirrorMaker2` 
+runtimes will be created with a single unencrypted listener on 8083 to match the existing behaviour.
+
+The `certChainAndKey` field will be required if the `https` protocol is selected. The `strimziTruststoreCert` will be required if the `https` protocol is selected and the provided 
```

Comment:
The HTTP server would only need a trust store if it was going to authenticate clients using their client certificates. IIUC this proposal is just for adding support for TLS encryption, so I don't think we need any trust store (perhaps listing authentication as a non goal would make sense). Instead the `certChainAndKey` will contain: i) The end-entity certificate of the server (with appropriate subject alternative name etc), ii) And needed intermediate CA certificates iii) The root CA certificate that REST clients will need to trust if they're going to be able to authenticate the server certificate.

It's then entirely up to the client to configure _its_ truststore with the CA certificate in the server's certificate chain. 

In the case where _we_ are managing connectors via `KafkaConnector` that client is in the CO, so we will have to add it to the truststore for the HTTP client we use. I suppose it would be fine to just do that automatically, using the `certChainAndKey` in the `KafkaConnect` resource, rather than having an extra place to configure this HTTP client (it doesn't seem to add any value that I can see).

Reply:
@tombentley I agree that Kafka Connect should not be creating a truststore, the truststore would only be used by the Strimzi operators. I put a question elsewhere about whether this would only be required if the managing via connectors is true. That was what I thought but wasn't sure if I'd missed something.

In terms of the certificates, the field I based it on is `brokerCertChainAndKey` and my reading of the docs "You can provide a certificate that includes the whole CA chain in the certificate file." is that you don't have to provide the whole chain, but I guess if we switched to requiring the chain for this field that would remove the need for the truststore field?

Reply:
It seemed a bit weird to configure the cert chain, complete with the root CA, and then, in the same resource, force the user to have to configure a trust store which contains that same root CA. But on thinking about it I suppose that (usually?) a cert chain doesn't include the root CA certificate, because it's assumed that the client will have it in its trust store. So maybe having a separate store for the root CA certificate makes sense for that reason alone. If you buy that argument, do you think we should validate that the cert chain _doesn't_ end with a self-signed CA cert (i.e. a root CA cert), to avoid misconfiguration?

> "You can provide a certificate that includes the whole CA chain in the certificate file."

I think this language is confusing. It's a bit ambiguous what a "certificate file" really is. I know what a _certificate_ is, and I know that I can store a sequence of them in PEM format in a file, and that would constitute a CA chain. I can also store certificate chains in other kinds of file (JKS, PKCS#12). But I don't think of any of those as a "certificate file", which sounds like it's only able to store a single certificate.

Reply:
If I understand correctly there are two possible ways we can go here:
1. The user only has to configure a `certChainAndKey` field, where they provide a complete certificate chain and the private key.
2. The user specifies two things, firstly `certAndKey` field where they provide the certificate that should be hosted by Kafka Connect and the private key. Secondly, only if they are using the KafkaConnector resource and their certificate is not signed by a well-known CA, they provide `truststoreCert` where they give the root CA certificate that the Strimzi Kafka Connect operator can trust when calling Kafka Connect

I think the way I wrote the proposal assumes option 2, but the naming of the cert and key field is then wrong. Which approach do you think makes more sense for the user @tombentley?

Reply:
As mentioned in a previous comment, we are heavily using `trustedCertificates` instead of `truststoreCert`.